Rsgraber/20230508 adr (#162)
openedx/event-bus-kafka: docs/decisions/0011_multiple_event_types_per_topic.rst

10. Multiple event types per topic
##################################

Status
******

> - **Accepted** DATE TO BE ADDED BEFORE MERGING

Reviewer: Nit: You can drop the `-`.

Author: [fixed]

Context
*******

Following `this ADR`_ in openedx-events, all implementations of event bus must support multiple event types on the same topic. This ADR explains how we will implement this using Kafka.
**UPDATE LINK AFTER THAT ADR MERGES**

> 

Reviewer: Can you add a small section providing a brief list of possible solutions to this problem in Kafka? Could just be bullets, and you can note that they'll be covered in more detail later. But basically, the context is that there are 2 (or 3, etc. depending on how you look at it) ways of accomplishing this, and this ADR is just about the one we selected.

Author: [fixed]

.. _this ADR: https://github.com/openedx/openedx-events/pull/217

Decision
********

> To comply with the event bus requirements, we will use the TopicRecordName strategy in the event serializers, which are responsible for registering and validating schemas against the schema registry.

Reviewer: Can you summarize why? For example: This method seems to be the simplest solution that solves the problem. (Or whatever you think the summary would be.) You don't really call out the pros/cons of the selected or rejected alternatives, so the why doesn't really jump out at you. Consider adding pros/cons, or reorganizing to make this more clear.

Consequences
************
* The record names for each event type will need to be distinct from each other. Originally, they were all just "CloudEvent." We needed to add namespaces to ensure that they were unique. Note that this will be necessary regardless of which solution we choose.
* Any event type will be able to be written to any topic, even for topics where we only want a single event type.
* Schema evolution rules will be enforced separately for each event type on a topic.
* The event-bus-kafka consumer will need to be updated to determine the signal from the message headers rather than taking a signal as a passed argument in the management command. This will also be necessary regardless of which solution we choose.

Rejected Alternatives
*********************

Setting the subject name strategy on the topic rather than the client
=====================================================================
This option is only available for locally-hosted or Dedicated Confluent Cloud clusters, and thus not a feasible solution for the whole community.

Using a SettingsToggle
======================

> Usually we would put this sort of feature behind a SettingsToggle so deployers could enable it and disable it as they chose. However, because the subject name strategy is set at the client rather than topic or cluster level, all clients who produce to the same topic(s) must have the same subject name strategy set. While the consequences of having multiple clients with different subject name strategies writing to the same topic are not 100% clear, at the very least it would result in a proliferation of confusing schemas in the registry. 

Reviewer: I'm not clear yet if this section will stay or not until we discuss on the other comment thread.

**IF** it stays, I wonder if we should:
- Add to the decision that we will **not** provide.a toggle to disable. (may not be necessary)
- Clarify that this is not an alternative to the TopicRecordName Strategy decision, but an implementation detail of that alternative. Maybe this would just move to the decision section altogether, and drop this from here?

Author: [fixed] removed

Using different serializers based on the topic
==============================================
We could refactor the code to configure the AvroSerializer differently based on the topic, only changing the subjec name strategy when the topic should allow multiple event types. This would be a significant amount of work and would require some list of which topics accept multiple event types, which would have to be maintained and would reduce the flexibility of the event bus.

Require related event types to have the same schema
===================================================

> In theory we could avoid the issue of event types with different schemas on the same topic by simply not allowing them, requiring all event types that are intended to go on the same topic to have the same schema. This would require knowing in advance which event types will go to which topics and likely result in lots of extraneous fields which are only necessary for some of the event types and not others.

Reviewer: Can you either drop this section, or just point to the other ADR that has the updated language?

Author: [fixed]

However, other types of events are not so easily grouped into a single signal. For example, there are different signals for ``XBLOCK_PUBLISHED``, ``XBLOCK_UPDATED``, and ``XBLOCK_DUPLICATED``, and indeed different schemas across the three signals. Routing these signals through different topics could lead to events being processed in a nonsensical order, for example an xblock being duplicated before it's been published. See `Should You Put Several Event Types in the Same Kafka Topic?`_ and `Putting Several Event Types in the Same Topic – Revisited`_ for more information on why it can be useful to group different event types on the same topic.

.. _Should You Put Several Event Types in the Same Kafka Topic?: https://www.confluent.io/blog/put-several-event-types-kafka-topic/
.. _Putting Several Event Types in the Same Topic – Revisited: https://www.confluent.io/blog/multiple-event-types-in-the-same-kafka-topic/

Decision
********

> We will update the code for event serializers, which are responsible for registering and validating schemas against the schema registry, to use the TopicRecordName strategy.

Reviewer: Maybe start with something like:

> We will allow multiple related event types to be produced to a single topic so that they can be ordered appropriately. We will accomplish this by...

Author: [resolved]

The default strategy the TopicName strategy. Under this strategy, the serializer will determine the existing schema by looking for a schema registered as ``<topic-name>-value`` (``<topic-name>-key`` when serializing the key), the topic name having been passed in the SerializationContext object. Therefore, if the serializer is given a new event type with the same topic, unless the new event type schema happens to be a valid evolution of the existing one, it will fail.

Using the TopicRecordName strategy, schemas are registered by topic and record name. On serializing an event, the serializer looks for a schema registered as ``<topic-name>-<record-name>``.
For example, given a schema like::

    {
        "doc": "Avro Event Format for CloudEvents created with openedx_events/schema",
        "fields": [{"name": "str_field", "type": "string"}]
        "name": "CloudEvent",
        "namespace": "my.event.type",
        "type": "record"
    }

> and the topic ``my-topic``, the serializer will look for a schema registered as ``my-topic-my.event.type.CloudEvent`` and use that to determine if the schema being passed is valid. This means that as long as the namespace is different for the schemas for different event types, the serializer will not error when adding new event types. It will still maintain evolution rules within an event type, however, which is desirable behavior.

Reviewer: It's unclear how exactly `record-name` relates to the `name` field and `namespace` field. Are these both standard fields looked at by something (what?), and is `namespace` used if provided, otherwise `name` is used?

Consequences
************
* The record names for each event type will need to be distinct from each other. Originally, they were all just "CloudEvent." We needed to add namespaces to ensure that they were unique. Note that this will be necessary regardless of which solution we choose.

> * Any event type will be able to be written to any topic

Reviewer: Maybe something like following, to make it clear that this is an unintended consequence?
```suggestion
* Any event type will be able to be written to any topic, even for topics where we only want a single event type.
```

> * Schema evolution rules will be enforced for event types but not for topics.

Reviewer: I wonder if "but not for topics" would be a little confusing for non-experts? What about something like:
```suggestion
* Schema evolution rules will be enforced separately for each event type on a topic.
```

* The event-bus-kafka consumer will need to be updated to determine the signal from the message headers rather than taking a signal as a passed argument in the management command. This will also be necessary regardless of which solution we choose.

> 

Reviewer: Somewhere could we note that in the future we might be able to detect if a service is only sending one event type for topic for all topics, and to not configure in the same way in this case?

Author: I'm not sure that's true. See the consequences of using a rollout toggle. We pretty much need this to be all-or-nothing across services.

Reviewer: Maybe we can have a chat, simply because I don't follow it. Only a single service should ever produce to a topic. I don't think we have an ADR around that, but that is a pattern that seems to be generally accepted. There might be several instances (e.g. EC2 instances, k8s containers) of the same service, but I don't think that is what you are referring to...

As I was writing this, I had a thought that _may_ be what you were trying to say. The above constraint is possibly about an event type, and not a topic. So, one question is whether this ADR should declare that all event types being produced to a shared topic must be sourced from the same service. I say yes.

Does this have anything to do with the point you were trying to make, or do we also need an entirely separate discussion as well?

Author: I was thinking about multiple services producing to the same topic. If we are comfortable saying that is not allowed and if you do that you could end up with weird schema problems, that's fine. I would definitely call that out in the ADR though.

Author: [fixed] Added back the SettingsToggle, and noted that this requires one producer per topic

Using a SettingsToggle
======================
Usually we would put this sort of feature behind a SettingsToggle so deployers could enable it and disable it as they chose. However, because the subject name strategy is set at the client rather than topic or cluster level, all clients who produce to the same topic(s) must have the same subject name strategy set. While the consequences of having multiple clients with different subject name strategies writing to the same topic are not 100% clear, at the very least it would result in a proliferation of confusing schemas in the registry. 

Using different serializers based on the topic
==============================================
We could refactor the code to configure the AvroSerializer differently based on the topic, only changing the subjec name strategy when the topic should allow multiple event types. This would be a significant amount of work and would require some list of which topics accept multiple event types, which would have to be maintained and would reduce the flexibility of the event bus.

Require related event types to have the same schema
===================================================

> In theory we could avoid the issue of event types with different schemas on the same topic by simply not allowing them, requiring all event types that are intended to go on the same topic to have the same schema. This would require knowing in advance which event types will go to which topics and likely result in lots of extraneous fields which are only necessary for some of the event types and not others.

Reviewer: Also, this is considered an anti-pattern, so we should definitely avoid this.

Author: [resolved]

10. Multiple event types per topic
##################################

Status
******

> **Accepted** DATE TO BE ADDED BEFORE MERGING

Reviewer: Reminder comment so this isn't forgotten.
